rfcs/syntax-string-interpolation.md

```diff
@@ -0,0 +1,122 @@
+
+The problems with `string.format` are many.
+
+1. Must be exact about the types and its corresponding value.
+2. Using `%d` is the idiomatic default for most people, but this loses precision.
+   * `%d` casts the number into `long long`, which has a lower max value than `double` and does not support decimals.
+   * `%f` by default will format to the millionths, e.g. `5.5` is `5.500000`.
+   * `%g` by default will format up to the hundred thousandths, e.g. `5.5` is `5.5` and `5.5312389` is `5.53123`. It will also convert the number to scientific notation when it encounters a number equal to or greater than 10^6.
+   * To not lose precision, you need to use `%s`, but even so the type checker assumes you actually wanted strings.
+3. No support for `boolean`. You must use `%s` **and** call `tostring`.
+4. No support for values implementing the `__tostring` metamethod.
+5. Having to use parentheses around string literals just to call a method of it.
+
+## Design
+
+To fix issue #5, we will need new syntax in order to not change the meaning of existing strings. There are a few components of this new expression:
+
+1. A string chunk (`` `...{ ``, `}...{`, and `` }...` ``) where `...` is a range of 0 to many characters.
+   * `%` escapes `%` and `{`, whereas `\` escapes `` ` ``.
+   * Restriction: the string interpolation literal must have at least one value to interpolate. We do not need 3 ways to express a single line string literal.
+2. An expression between the braces. This is the value that will be interpolated into the string.
+3. Formatting style may follow the expression, delimited by `,`.
+   * Restriction: the formatting style must be constant at parse time.
+   * The formatting style will follow the same syntax as pre-existing syntax in `string.format`.
+
+To put the above into formal EBNF grammar:
+
+```
+fmtflags ::= `-' | `+' | ` ' | `#' | `0'
+fmtwidth ::= <NUMBER> [<NUMBER>]
+fmtprecision ::= `.' <NUMBER> [<NUMBER>]
+fmtspecifier ::= `c' | `d' | `i' | `e' | `E' | `f' | `g' | `G' | `o' | `q' | `s' | `u' | `x' | `X'
+
+interpfmt ::= [fmtflags][fmtwidth][fmtprecision][fmtspecifier]
+
+interpexp ::= exp [`,' interpfmt]
+
+stringinterp ::= <INTERP_BEGIN> interpfmt {<INTERP_MID> interpexp} <INTERP_END>
+```
+
+Which, in actual Luau code, will look like the following:
+
+```
+local world = "world"
+print(`Hello {world}!`)
+--> Hello world!
+
+print(`0x{3735928559,X}`)
+--> 0xDEADBEEF
+
+local combo = {5, 2, 8, 9}
+print(`The lock combinations are: {table.concat(combo, ", ")}`)
+--> The lock combinations are: 5, 2, 8, 9
+```
+
+To fix issues #1, #2, #3, and #4, we will also want to extend `string.format` to accept braces after `%`. This is currently an invalid token, so this is a backward compatible extension. Unfortunately, formatting style cannot immediately follow `%` because `%d{}` is already valid and parses as `(%d){}`.
+
+The extended syntax will allow the following ``%{[<NUMBER>][`,' interpfmt]}`` to be parsed as one token.
```

Comment:
I don't fully understand why we need to add number specification to string.format. It doesn't seem necessary for string interpolation. It feels that for string interpolation the minimal required change to string.format would be to add a format specifier that formats the argument without requiring it be a specific type, and everything else can just be part of the interpolated strings?

Reply:
In general I think we should reframe this RFC as "String interpolation", and only have the minimal extensions to string.format that support that. We should argue for why string interpolation is better than manual concatenation or string.format, and then talk about the lowering of interpolated strings into string.format and *that's* where string.format extensions come in. That helps minimize the cost of this feature.

Reply:
Sounds good. The thing that I was afraid about is the inability to extend it if we went with the wrong design to start, which is why I initially went for a more complete RFC.

For example, leaving out the explicit positional numbers while keeping in formatting specifiers might be ambiguous if we want to have both, and we didn't have a required delimiter. Is `%{9}` supposed to mean the 9th argument or right-fill with up to 9 spaces? Explicit positional numbers could be useful if you have a function that returns multiple things and you'd like to reorder it in a different way. I suppose intermediate locals will help and improves readability.

Anyway, I'll cut down the RFC to focus specifically on string interpolation and also try to keep `string.format` open for future extensions. I suspect that at this time we probably don't need any of the formatting specifiers as part of the syntax just yet either. So we'll leave those as undefined in this RFC.

Reply:
Since `%{` is not valid today in `string.format` if I remember correctly, as long as we don't use it we still leave the format open for this extension in the future, and I *think* we don't need it for string interpolation even if the interpolation itself uses `{}` as a delimeter.